### PR TITLE
[v7r0] Backport GitHub actions configuration

### PR DESCRIPTION
**Pull request by chrisburr**
Backport #4289

BEGINRELEASENOTES
*Core
NEW/CHANGE/FIX: explanation
FIX: Use subprocess32 if possible for significantly better performance
FIX: TypeLoader should not be a singleton as it has a cache
*FrameworkSystem
FIX: Avoid leaking std(in|out|err) file handles when running runsvdir
*Tests
NEW: CI for unit/lint/integration testing with GitHub actions
NEW: Start using shell linting of the test scripts
FIX: Exit quickly if the installation fails to make it easier to find issues
*Doc
NEW: Document how to run integration tests in docker
*Misc
NEW: Add environment.yml file for preparing an environment with conda
ENDRELEASENOTES

### CODE REVIEW

**Comment by atsareg** [comment]
Is it normal that integration tests fail in this case ?

**Comment by chrisburr** [comment]
No it isn't. There must be something else in the test code that needs to be backported/fixed but I won't have time to look at it today.

#4286 can still be merged to add tests to the integration branch.

**Comment by atsareg** [comment]
Am I understanding correctly that #4286 is the same functionality as #4289 but for the integration branch ? If this is the case, then this functionality should be added to the integration branch by merging the rel-v7r1 to integration and not by making a special PR.

**Comment by fstagni** [comment]
I would not change `subprocess` to `subprocess32` in this branch, which is already a semi-production branch. In general, I would not modify DIRAC .py files here, only add/modify the test part.

File: .travis.yml [hunk 1]
```diff
@@ -30,26 +30,33 @@ install:
...
+    elif [[ "${CHECK}" == "shellcheck"  ]]; then
+      # TODO This should cover more than just tests/CI
+      find tests/CI -name '*.sh' -print0 | xargs -0 -n1 shellcheck --external-sources
+    elif [[ "${CHECK}" == "pylint"  ]]; then
+      travis_wait 30 .travis.d/runPylint.sh;
+    elif [[ "${CHECK}" == "pylintPY3K"  ]]; then
+      travis_wait 30 .travis.d/runPylint.sh;
+    elif [[ "${CHECK}" == "format" ]] && [[ "${TRAVIS_PULL_REQUEST}" != "false" ]]; then
+      git remote add GH https://github.com/DIRACGrid/DIRAC.git; git fetch --no-tags GH ${TRAVIS_BRANCH}; git branch -vv; git diff -U0 GH/${TRAVIS_BRANCH} ':(exclude)tests/formatting/pep8_bad.py' | pycodestyle --diff;
+    else
```
Comment (andresailer) [inline .travis.yml]: This leads to non-PR test runs failing for the "format" check. As is the case for the current integration tests.

File: environment.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  - suds-jurko >=0.6
+  # testing and development
+  - autopep8
+  - caniusepython3
+  - codecov
+  - coverage
+  - hypothesis
+  - ipython
+  - mock
+  - pylint >=1.6.5
```
Comment (fstagni) [inline environment.yml]: You need to add `parametrized` here (see e.g. https://github.com/DIRACGrid/DIRAC/pull/4103/checks?check_run_id=263589384#step:4:23) which was added to `requirements.txt` from rel-v6r22

**Comment by chrisburr** [comment]
> I would not change `subprocess` to `subprocess32` in this branch, which is already a semi-production branch. In general, I would not modify DIRAC .py files here, only add/modify the test part.

@fstagni Without changing `subprocess32` the tests run too slowly to be usable (~2 hours). I think it's safe enough to to not be worth the effort of selectively removing.

File: .travis.yml (hunk 1 above)
Comment (chrisburr) [reply .travis.yml]: Was fixed in #4350.

**Comment by fstagni** [comment]
> > I would not change `subprocess` to `subprocess32` in this branch, which is already a semi-production branch. In general, I would not modify DIRAC .py files here, only add/modify the test part.
> 
> @fstagni Without changing `subprocess32` the tests run too slowly to be usable (~2 hours). I think it's safe enough to to not be worth the effort of selectively removing.

Alright, let's try, but this is an exception to what we would normally do. We'll anyway verify this release in a LHCbDIRAC certif hackathon next week.

**Comment by chrisburr** [comment]
Okay I've tried merging this on on my fork and then propagating it to `rel-v7r1` and `integration`. I then merged #4348 into `rel-v6r22` and propagated that without any issues.

To make this work I had to:
- Include aedaf7e to make `dirac-install` support installing from a directory
- Merge #4314 **FIRST** to avoid messy conflicts when propagating this forwards.

**Comment by chaen** [comment]
So, big merging campaing now ?

**Comment by chrisburr** [comment]
The test that I changed in 797e5a7 has already passed so it's safe to merge without waiting for the others

**Comment by fstagni** [comment]
Please just resolve the conflict since you are on it

**Comment by chrisburr** [comment]
Done and the CI has passed

**Pull request merged**